### ISSUE TITLE
Zip/7z mount and launch

### ISSUE THREAD

**Issue by enderandrew** [dosbox-staging/dosbox-staging#314]
I prefer to keep my collection of old DOS games in zip files rather than thousands and thousands of loose files. The old DAUM fork allowed for mounting of a zip file or a 7zip file as a directory.

That had to be done via a config file, though ideally a newer implementation would allow dropping a zip file onto the DOSBOX executable, launching via a command line switch, and/or selecting a zip file from a DOSBOX menu that would mount the archive as a folder and perhaps even look for a batch file in the root of that folder structure and launch it.

Per-game configuration that you modify in DOSBOX while running the game could perhaps be rewritten back into the zip file so that the next time you launch that game from the zip file again, there would be persistence of your configuration. (Conversely you could save those configurations in your user profile folder, but looking for a config file in the game archive is a nice way to assist with distributing proper configurations with the games themselves).

**Comment by dreamer** [comment]
This is a feature, that we touched upon several times with @kcgen - some preliminary work landed already (#143), other pieces of the puzzle were investigated with initial implementation (#92). We *definitely* want to go in this direction, but we have a different view on the implementation details/scope of the feature.

Basically, there are several *pretty important* use cases, that would be broken with a simple compressed file with game + .conf inside (examples: archiving games with compressed audio would be hampered by putting a compressed file into a compressed archive, a mechanism for handling file writes would be very implementation-specific if we would go without a spec, there are compression algorithms better suited than zip/7zip performance-wise, people would definitely want to store metadata with the game, so it opens the field for competing schemes for storing meta, etc, etc).

This topic keeps coming back, but users requesting it often put their own spin on the feature tailored to their particular homegrown solution to the problem: (as was seen in discussions for #99 #166 #260)

The way I see it: in this case, we shouldn't rush to implementation (but we will continue with implementing some "pieces of the puzzle" as preparatory work), but rather start by creating a proper RFC document specifying new container format (ROM-like files for DOS games - my idea for naming it was `.dsc` for "DOS Software Container") - and with RFC document to show, ask for feedback from users, developers, other dosbox forks, dosbox frontend developers, game archivists, etc.

Once I'll have a while, I will prepare a feature request specifying exact scope of the feature, that we want to implement and some technical details - @enderandrew after that plan will be in place I'll ask for your feedback on the feature scope and perhaps close this issue (alongside other feature request issues I listed above), so we will continue discussion in a single place.

**Comment by enderandrew** [comment]
There are certainly multiple ways to skin a cat and I'm not picky.

I will say that if you basically create a DOS rom format, you have to convince everyone to convert their existing libraries, and you have to convince everyone they should want to use your format in the first place. This can lead to others spinning up competing formats at once.

I assume most DOS games are currently archived in ZIP files currently (though some just archive the installer floppy discs or CD roms, and others are zipping up the extracted/installed contents that you can more readily just run right away).

It might make sense to support multiple methods. Propose your DOS rom format, but also allow people to drop their existing zip files and attempt to launch a bat file if you see one, or an executable if there is a single executable in the root of the zip file. (If there are multiple executables I'm not sure if there is a good way to determine the main executable).

Again, I'm not picky, that is just my two cents.

**Comment by enderandrew** [comment]
If you do propose a rom format, it might be nice to create a tool to try and convert existing folders or zip files to the proposed rom format.

**Comment by dreamer** [comment]
Except there is no unifying standard nor described format for packaged zip files, I know about at least 3-4 different ones (I don't know how incompatible they are with each other).

That's why I want to start with RFC - to learn what usecases people have in mind (archival usecase with cue/bin, is different than personal archive with cue/opus, games with multiple CDs, or game distributor's usecase, or demoscene usecase, etc, etc) - all those usecases are not competing with each other and can share the same container format - but they put emphasis on a different aspects of packaging.

If 2 dosbox forks agreed on the format and implemented it - I guess that would make it pretty clear what to use throughout the ecosystem :)

As for zip - my opinion is that `zstd` algorithm is much better suited in every way for the purpose (at least Fedora packaging performance tests indicate so), it's also widely available on every modern OS nowadays AFAIK.

Creating reference implementations for tools used for packaging is a critical part of this work.

**Comment by theMK2k** [comment]
I second the feature request. Please add 7z/zip mount.

Please keep in mind though that the DAUM 7z/zip mount also allowed to add another layer for the mount. That layer is a (empty) directory. Now if anything is written on the mount point (e.g. a save game, game config etc.), the change would be written to the (empty) directory.

Upon the next start, the combination of 7z/zip contents and the contents of the (now not empty) directory are merged and the changed content is read when accessed.

In short: we're not really talking about ROM as writes are theoretically supported.

**Comment by dreamer** [comment]
@theMK2k 
Oh, yes, of course - ROM file is only a way to archive/distribute the game - all writes (save games, game settings, etc) will be allowed and redirected to a separate directory (there was a problem with DOS games opening game resources in write-mode, even when they did no writes - we fixed it in #143).

This feature ("mount overlay") is already supported - SVN merged change submitted by GOG team ~1 year ago - it's undocumented and was never advertised AFAIK, but you can already use it in 0.75.0-rc1:

    mount C <path> -t overlay

I will describe it in the changelog for 0.75.0 final release, as it's a pretty important feature :)

**Comment by i30817** [comment]
Does overlay work with hdd images, specifically 'raw' mounted images (images mounted with numbers, not drive letters, for windows to manage later?). My external linux copy on write mount handles this by copying the whole hdd file, which is probably the only option for FAT32 because the dosbos shell pretends it's opaque when windows 95/98 boot it so i was wondering if the overlay behaves in this case (or even works better by copying only the altered files inside the hdd, which i doubt).

**Comment by dreamer** [comment]
@i30817 For directory-based mounts, overlays only copy the altered files into the overlay dir.
AFAIK, overlay mounts are not compatible with drives mounted via `imgmount` yet.

If you want something like this working, then make sure it does *not* work for your use-case (I don't know how do you mount your images and what type of disk images exactly are you referring to), then create a feature request exactly explaining your use-case. I've shown example overlay mount command in my previous post. Thanks in advance.

**Comment by i30817** [comment]
> AFAIK, overlay mounts are not compatible with drives mounted via imgmount yet

This is what i was looking for. Filesystem OS VFS superiority remains i guess (i have some windows 95 games running in dosbox that need mounts of hdd image files running in read only dirs).

**Comment by dreamer** [comment]
@i30817 FYI: Yesterday I talked to certain prominent DOS game archive owner about this, and overlay+imgmount seems to be a high-priority feature request for specific use-cases. ;)

Ability to mount directory overlays on top of read-only imgmount mounts would also serve as a stepping stone towards *this* feature request.

**Comment by i30817** [comment]
In unixes there is no difference between a read only mount and a zip mount. It's the only sane way to do things (though of course, many programs do insane things in the name of windows compatibility because windows didn't have a userspace VFS api until recently).

**Comment by Akuma-Git** [comment]
> If you do propose a rom format, it might be nice to create a tool to try and convert existing folders or zip files to the proposed rom format.

> @theMK2k
> Oh, yes, of course - ROM file is only a way to archive/distribute the game - all writes (save > 

The standard afaik is 'read-only zip' files, if a new ROM-format is introduced adoption will be low. There are a couple of reasons that come to mind:
* SVN needs to adopt this for it to get any traction outside staging
* Most do not want to convert their existing files / collections.
* Consensus on a ROM format between projects.

But ZIP support itself should be a common denominator.

Now I happened to stumble on 'fuse-zip' (LGPL) which works afaict pretty good (need some more testing) and is fast.

@dreamer @kcgen : https://github.com/ralic/fuse-zip

**Comment by i30817** [comment]
Zip is all things considered... pretty bad for a rom format. There is worse, like rar, but zip has no programmable metadata at all and that inevitably degenerates into 'let's put metadata into the filename'.

It also has no way to attach metadata to the format without changing the external checksum, which means that whenever you try to add it, you inevitably ruined your database, or you scan internally all zips (or whatever extension you use to pretend).

**Comment by dreamer** [comment]
@Akuma-Git 

> The standard afaik is 'read-only zip' files

There is no standard when it comes to DOS games distribution. There were few attempts in the past (e.g. in Daum), but zipping a bunch of files together is *NOT* the solution.

> SVN needs to adopt this for it to get any traction outside staging

SVN will do nothing - they don't accept even crash fixes from us. They are not going to accept any new features.

> Most do not want to convert their existing files / collections.

I don't care - we are doing it for ourselves (just like *all* the work we do in dosbox-staging). If the solution will be good enough and tools to assist with migration will be there, then users/archives will start switching. We already talked to some key community figures and we know what other projects are thinking, have somewhat good idea about the requirements.

> Consensus on a ROM format between projects.

There are plenty of DOSBox forks, but there are only 2 that are actively maintained and willing to implement features of this calibre: dosbox-staging and DOSBox-X. Building consensus depends on having RFC document first, before we jump off with implementation.

If other projects will just implement yet another stupid zip-based solution, then we simply won't adopt it, because *it won't work for us*.

> But ZIP support itself should be a common denominator.

No, it won't be. ZIP files will provide non-optimal compression ratio, too slow decompression speeds, and will not work for game packages that use compressed CD audio.

> fuse-zip

Won't work on Windows, macOS, nor BSDs - therefore it's not an acceptable solution. If we were to support Linux only, this feature would be rather easy to implement. But to have any impact on the community at large, we need it to be a cross-platform solution.

> Use chd. This case is literally what it's for. 

No, it's not. I skimmed through chd reference implementation and "spec" and it seems to be missing a number of features that we need - it's ~70% there, so extending chd *might* be an approach we will take. We'll see.

**Comment by i30817** [comment]
What do you think is missing for chd? I has the single file internal checksum, it's agnostic to cd vs hard-drive sectors vs files (i think), it has a softpatching mechanism that ignores the compression, it has metadata that doesn't interfere with the internal checksum?

What's missing except application concerns like configurations and hardware descriptions?

**Comment by theMK2k** [comment]
Please don't neglect .zip as it is also supported on DOS hardware (e.g. https://github.com/MobyGamer/total-dos-launcher), this way a game collection can be used in modern OS with dosbox-staging as well as vintage scenarios.

**Comment by i30817** [comment]
Hell. You aren't going to get multigb games running with zip on 'vintage scenarios'. If that thing is forced to update to support a saner format it's all to the good for them, if they don't, convert the games yourself and suffer the consequences of very dump 'extract everything everytime i want to play a game'.

I swear some pc users upgraded to 16/32 gb of memory lately and are going mad with power because they can place gb files into memory. Not on your DOS machine you aren't.

**Comment by i30817** [comment]
edit: removed kind of mean post, but suffice to say i don't agree it's good to sacrifice performance and disk lifetime and space because some programs just want to uncompress zips.

**Comment by dreamer** [comment]
*I freely admit, that I might be wrong, maybe chd format / reference implementation changed since last time I looked at it ~9 months ago.*

@i30817 
1. To the best of my knowledge, CHD deals with a single "image" (consisting of multiple hunks; different hunks might use different compression as of v5) - this is ok for all consoles, cabinets, etc - but we need a mix of several "images" per ROM file. Usecases we need to support are e.g.:
    - package several floppy images (up to ~50) + installed data + metadata
    - several CD images + installed data + metadata
    - mix of CD images and audio files + cues + installed data + metadata
    - raw disk images + metadata
    - (probably other variations and combinations as well)

2. CHD is missing the support for compressing hunks using algorithms we need: zstd and xz
3. We can't have an unspecified blob of metadata - we need to have it structured and specified which emulator config options are allowed to be embedded in there (example: `render.aspect` belongs to metadata, `sdl.output` does not). We also need structure, so that ROMs won't need to be modified for different frontends.
4. ROM files need to be, well, ROM (no write backs), which means we need automatic overlay setup (+path filters to avoid buggy games "leaking" totally out of ROM image into the overlay). This is really out of scope of CHD.
5. (not for v1) We need content negotiation algorithm - e.g. game package can say: "I want to play music using MIDI, if not available then MT-32, if not available then SoundBlaster" and emulator should be able to pick the settings that are supported, working (e.g. user provided ROMs or SF2s), and user-preferred, *and trigger game reconfiguration to pick negotiated settings*. We need this to be part of specification, otherwise ROMs won't work correctly.
6. I *suspect* CHD does not support padding on the image boundaries, which I *think* will be necessary to effectively memory map different parts of the ROM file into the memory on demand. As all previous points, I might be wrong on this, haven't dived into the reference implementation so deep.

If we'll go the route of extending CHD (which I would say is likely), then of course we'll talk to MAME devs and coordinate :) But as for now, I am not actively working on this yet - still gathering usecases and working on stuff, that is more critical to our users.

> I swear some pc users that are programmers upgraded to 16/32 gb of memory lately and are going mad with power because they can place gb files into memory. Not on your DOS machine you aren't.

Yup. Even on modern machines - we simply **cannot** have a ROM file consisting of several CD images loaded into the memory all at once. We'll need to have dosbox mmap only the required parts of ROM file (so when CD swap occurs, old image/audio files are freed and only new ones are loaded into the memory). This is crucial for e.g. Raspberri Pi users (which are more numerous than vintage hardware gamers anyway), or for future Android port.

**Comment by kcgen** [comment]
CHD's interface is at the block level.. hence "hunks of data", on-top of which the developer places the ROM structure or more complex formats, like a filesystem.  

So the API into it isn't nearly as elegant as something like PhsFS, which operates like its own filesystem. "Give me a list of files in directory X"), for example.

CHD doesn't have threaded read-ahead decode (which is also true for all the fuses).  Yes, the host OS will read-ahead the next underlying block that underpins your CHD file or ZIP file, but there's 1 no guarantee that specific block corresponds to the next block of data in the logical file you're working with inside dosbox and 2 the host's readahead only means the underlying physical IO has been performed ahead of time: the logical handling of that block still needs to performed by your compressed format (such as crc-check and decompression), so for CHD, that inserts a non-trivial gap of latency inside your game-loop.

The only way to work around this (if one decides to use CHD), is to use a small-enough compressed block size _for the platform you're running on_ that the decode latency is fast enough that it doesn't harm gameplay. 

That all sounds theoretical, but lets use a real-world example: PlayStation 1 emulator,  CHD ROM format using 1 MiB LZMA blocks, running on a 4-core 1.4 GHz Raspberry Pi 3.   

You will experience stuttering during 3D + IO intensive streaming sequences, such as is common in the game RayCrisis (many 3D PS1 games have these periods, but the Ray-games are particularly full of them). This is because injecting the latency of decompressing 1MiB LZMA blocks "just in time" inline into the emulator's loop adds way too much latency on the Pi3. 

The only solution for PS1 + CHD kn the Pi3 is to use zlib compression at it's default block size, or LZMA with 256KiB or smaller blocks.

Threaded "next logical content" decode will zero out that latency so the game itself is handed the data with near zero latency (similar to how a physical CDROM drive works). 

And on top of it, we will use best in class compression (such as zstandard) - which can decode between 10x and even 30x quicker than LZMA, with little change in decompression speed versus compression level and block-sized used.  So our readahead decode thread will also be light weight (given heavier alternatives). 

This is just one aspect I'm picking on here.. but it's fundamental.

**Comment by Akuma-Git** [comment]
> @Akuma-Git
> 
> > The standard afaik is 'read-only zip' files
> 
> There is no standard when it comes to DOS games distribution. There were few attempts in the past (e.g. in Daum), but zipping a bunch of files together is _NOT_ the solution.
> 

'zip' is what most users have, so you could go for an automatic or intermediate solution.
On-the-fly conversion could bring both worlds together. Otherwise imo, you will not see widespread adoption.

> > SVN needs to adopt this for it to get any traction outside staging
> 
> SVN will do nothing - they don't accept even crash fixes from us. They are not going to accept any new features.

Yes, that is unfortunate and does not help I agree.

> > Most do not want to convert their existing files / collections.
> I don't care - we are doing it for ourselves (just like _all_ the work we do in dosbox-staging). If the solution will be good enough and tools to assist with migration will be there, then users/archives will start switching. We already talked to some key community figures and we know what other projects are thinking, have somewhat good idea about the requirements.

I get that, but I disagree on this one, but I like to be proven wrong.

> > But ZIP support itself should be a common denominator.
> 
> No, it won't be. ZIP files will provide non-optimal compression ratio, too slow decompression speeds, and will not work for game packages that use compressed CD audio.

Hmm, it works fine with the zips I tested containing FLAC/OPUS
Tests done from a:
* 2GB USB2.0 thumb-drive.
* 10 speed micro sdcard.

> > fuse-zip
> 
> Won't work on Windows, macOS, nor BSDs - therefore it's not an acceptable solution. If we were to support Linux only, this feature would be rather easy to implement. But to have any impact on the community at large, we need it to be a cross-platform solution.

that is unfortunate :(

**Comment by i30817** [comment]
Why would you want to package several cds or floppies in the same container?

Let me honest here: multiple 'physical media' containers are terrible. The only example that exists, pbp sucks (mostly because the projects implementing it invariably screw up the 'multiple' part), and the situation is much much worse for DOS games.

**Comment by i30817** [comment]
That said, if i could, i'd have all games in a set be single layer, filesystem transparent like a zip, and without need for multiple devices (like a full install). That's the ideal scenario for a copy on write system on dosbox, which has filesystem transparency.

The container would need some kind of metadata, and amortized unique ID not affected by embedded metadata. Configuration being part of metadata (and thus affecting the ID) is a question of taste, but i'm tending to 'not' to make 'canonical sets' using 'dumb checksums' possible.

**Comment by dreamer** [comment]
I encourage people who have strong opinions about how the solution should work to start working on a solution.

**Comment by nemo93** [comment]
Closing this one as another ticket has been raised to discuss this ROM package specification. #939

**Comment by Torinde** [comment]
Related:
- mounting ZIP/7Z archives as drives and run games in these mounted drives directly. They will operate in read+write mode with overlay mounts.
  - https://github.com/joncampbell123/dosbox-x/issues/1132
  - https://github.com/joncampbell123/dosbox-x/issues/3535
- [.DOSZ](https://github.com/schellingb/dosbox-pure?tab=readme-ov-file#zip-files-can-be-renamed-to-dosz) (ZIP extension renamed for Pure), [.D.ZIP](https://github.com/schellingb/dosbox-pure?tab=readme-ov-file#mount-zip-as-a-or-d-drive) (drive letter to be inferred from filename)

**Comment by johnnovak** [comment]
I don't think this will happen. Kinda feature bloat for no good reason, plus a read-only C drive is just silly.

**Comment by eduo** [comment]
I was thinking of trying my hand at this for when CDs and Floppies are mounted from directories (something common in some game distributions out there and also common with people who modify and hack game files).

That is, as an addition to imgmount, rather than to mount. Or to mount but explicitly to -t cdrom.

**Comment by johnnovak** [comment]
> I was thinking of trying my hand at this for when CDs and Floppies are mounted from directories (something common in some game distributions out there and also common with people who modify and hack game files).
> That is, as an addition to imgmount, rather than to mount. Or to mount but explicitly to -t cdrom.

What do you mean exactly? A ZIP containing the contents of the CD? I don't really see the point. First, storage is cheap, so just unzip it and mount the directory normally. Second, CUE/BIN and ISO is strongly preferred.

So I'm not seeing why this would be worth the significant extra complexity.

If the ZIP uses compression level 0 (store only), we could just seek into it, ok. But if it's a 200-500 MB or larger ZIP file, do we really want to to decompress the whole thing always on the fly? Then store it in memory? That would be super slow. I can't see anyone wanting to do this and pay the startup delay price every single time. Just archive the ZIP file somewhere if you wanna keep it for archival purposes, then unzip it for actual use.

It's a lot of complexity for some slim benefits (maybe) for some users, but the drawbacks are significant.

**Comment by eduo** [comment]
You're right. I was thinking of specific use cases where people had CDROMs as individual files mounted via mount but thinking it better it would be terrible either for memory (keep the whole thing in memory) and speed (uncompress to begin with) or would be terrible for performance (if uncompressed on demand, when there are frequent reads).

I was thinking it could be treated as an ISO, but ZIP is really not optimized for that usage.
